### ISSUE TITLE
Optimize Cannon memory merkleization

### ISSUE THREAD

**Issue by mbaxter** [ethereum-optimism/optimism#11695]
## Description

Cannon's memory is currently structured as a binary merkle tree of depth 27 with 32 byte leaf values, representing a 32-bit address space.  There are some performance bottlenecks with the current implementation.  In particular, [cache invalidation](https://github.com/ethereum-optimism/optimism/blob/d520441b008833d8ee8ee69bbe2e8158366279d0/cannon/mipsevm/memory/memory.go#L202-L202) and [merkleization](https://github.com/ethereum-optimism/optimism/blob/d520441b008833d8ee8ee69bbe2e8158366279d0/cannon/mipsevm/memory/memory.go#L134-L141) are expensive. Cache invalidation touches a lot of memory and we have to walk down the tree invalidating cached nodes one by one.  Similarly, we do a lot of hashing when remerkleizing as data is updated. 

Some preliminary prototyping suggests we could get better performance by moving to a radix trie like the [modified Merkle-Patricia Trie (MPT)](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/) used to store Ethereum state data.  Radix tries can more efficiently store sparse data, minimizing time spent looking up values and reducing merkleization costs.  A benefit of MPTs is that we already have a [Merkle-Patricia Trie solidity implementation](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L330-L335), which should simplify on-chain integration.  

We can start experimenting by implementing a new Memory format using a MPT.  To avoid committing to the new format, we can create a [Memory interface](https://github.com/ethereum-optimism/optimism/blob/d520441b008833d8ee8ee69bbe2e8158366279d0/cannon/mipsevm/iface.go#L11-L11) so that we can swap different memory implementations into the VM while testing.  We'll want to start by benchmarking off-chain performance before thinking about fully integrating any new approach.

During preliminary prototyping with a MPT, we used leaf values of 4096 bytes ([1 page ](https://github.com/ethereum-optimism/optimism/blob/d520441b008833d8ee8ee69bbe2e8158366279d0/cannon/mipsevm/memory/page.go#L23)in the current implementation), with 20-bit keys. We may want to start with an off-the-shelf implementation to see what performance looks like.  We could consider implementing our own MPT if we can't find a sufficiently performant implementation.  We could also consider experimenting with the structure of the trie in other ways, like modifying the branching factor (standard MPTs are hexary - branches have 16 children), though it may be simpler to stick with a standard format.

**Comment by zhiqiangxu** [comment]
> Radix tries can more efficiently store sparse data, minimizing time spent looking up values and reducing merkleization costs. 

Is the benefit mainly a result of the reduction of depth of the tree, or in other words, from binary tree to n-ary tree?

**Comment by mbaxter** [comment]
Yeah, I think the higher branching factor helps.  It means fewer nodes to visit during lookups and fewer nodes to rehash when a value is updated.  With MPTs you also have prefix collapsing where you can have branch or extension nodes that store a common path prefix for their children, which also reduces tree depth.

@clabby - anything to add here?

**Comment by ec2** [comment]
Howdy! So I've been working on this a bit and just wanted to give a little progress report.

So far, I've:
- abstracted the Memory struct into a now Memory interface.
- implemented the Memory interface where the page tables are backed by
   - the current binary tree approach
   - copy pasta of asterisc's MPT approach

I will clean up some stuff and push a draft PR, but just wanna give an update since I'm excited lol

**Comment by ec2** [comment]
Is there a a reason why we should be representing pages as [PageSize]byte{}? It seems like we should really be storing them as [PageSize/(WordSize/8)]Word instead.

Right now, every memory access involves endian conversions. But if we store them as Words, then we only need to do endian conversions when hashing (which happens way less often than memory accesses). 

This change is easy assuming we continue only disallowing unaligned memory accesses.
